---
title: "GECs methylation profiles"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{hydromad}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r echo=FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide")
source("common.R")
```
```{r params}
tcga_project = "TCGA-BRCA"
gecs = c(
  "CENPE", 
  "CENPF", 
  "DNMT3B", 
  "EXO1", 
  "MCM10"
)
s_cnv   = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_cnv.rds"))
s_trscr = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_trscr.rds"))    
gene_filename="~/projects/genes/bed_grch38_epimeddb.rds"
genes = mreadRDS(gene_filename)
```

# Genes

```{r, results="verbatim"}
genes[gecs, c("chrom_text", "tx_start", "tx_end", "tss_cpg_status")]
```


# Expression

```{r}
gene_list = openxlsx::read.xlsx("../data/Gene_list.xlsx")
gene_list
exp_grp = openxlsx::read.xlsx("../data/TCGA-BRCA_sample_annotations.xlsx")
rownames(exp_grp) = exp_grp$id_sample
dim(exp_grp)
```

```{r}
layout(matrix(1:2, 1), respect=TRUE)
for (g in gecs) {
  print(g)
  den = density(s_trscr$data[g,])
  bw = den$bw
  plot(den, main=g)

  idx_normal = rownames(exp_grp)[exp_grp$tissue_status=="normal"]
  idx_normal = intersect(colnames(s_trscr$data), idx_normal)
  idx_off = rownames(exp_grp)[exp_grp[, paste0(g, "_off_on")]%in%0]
  idx_off = intersect(colnames(s_trscr$data), idx_off)
  idx_on = rownames(exp_grp)[exp_grp[, paste0(g, "_off_on")]%in%1]
  idx_on = intersect(colnames(s_trscr$data), idx_on)

  den_normal = density(s_trscr$data[g,idx_normal], bw=bw)
  lines(den_normal$x, length(idx_normal) / ncol(s_trscr$data) * den_normal$y, main=g, col=3)

  den_off = density(s_trscr$data[g,idx_off], bw=bw)
  lines(den_off$x, length(idx_off) / ncol(s_trscr$data) * den_off$y, main=g, col=4)

  den_on = density(s_trscr$data[g,idx_on], bw=bw)
  lines(den_on$x, length(idx_on) / ncol(s_trscr$data) * den_on$y, main=g, col=2)
  legend("topleft", c("normal", "OFF", "ON"), col=c(3, 4, 2), lty=1)
}
  

```

# Methylation of promoters

```{r}
s = mreadRDS("~/projects/tcga_studies/study_TCGA-BRCA_meth.rds")


get_genes = function(tcga_project, gene_filename="~/projects/genes/bed_grch38_epimeddb.rds") {
  genes = mreadRDS(gene_filename)
  if (!missing(tcga_project)) {
    s_cnv   = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_cnv.rds"))
    s_trscr = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_trscr.rds"))    
    genes = genes[rownames(genes) %in% intersect(rownames(s_trscr$data), rownames(s_cnv$data)),]      
  }
  ## index meth probes by chr
  pf_chr_colname = colnames(genes)[1]
  chrs = as.character(unique(genes[[pf_chr_colname]]))
  chrs_indexed_genes = lapply(chrs, function(chr) {
    # print(chr)
    idx = rownames(genes)[!is.na(genes[[pf_chr_colname]]) & genes[[pf_chr_colname]]==chr]  
    ret = genes[idx,]
    return(ret)
  })
  names(chrs_indexed_genes) = chrs

  return(list(genes=genes, chrs_indexed_genes=chrs_indexed_genes))
}

#' get_probe_names
#'
#' This function extracts probe names of a given gene from platform
#' @param gene A vector describing the gene (line of a bed file).
#' @param pf_meth A data frame describing CpG positions.
#' @param up_str   An integer specifying up stream size (in bp).
#' @param dwn_str  An integer specifying down stream size (in bp).
#' @param pf_chr_colname string matching the name of the column in the platform that contain the chromosome on which we find a probes.
#' @param pf_pos_colname string matching the name of the column in the platform that contain the position information of probes.
#' @return A vector of probe names
#' @export
get_probe_names = function(
  gene                                ,
  pf_meth                       , 
  pf_chr_colname="Chromosome"         ,
  pf_pos_colname="Start"              ,
  up_str=5000                         , 
  dwn_str=5000                        
 ) {  
   if (substr(gene[[1]], 1, 3) != "chr") {
     gene[[1]] = paste0("chr",gene[[1]])
   }
    
   # get gene properties
   chr =            gene[[1]]
   strand =         gene[[6]]
   gene_name =      gene[[4]]
   beg = as.numeric(gene[[2]])
   end = as.numeric(gene[[3]])

  if (nrow(pf_meth) == 0) {
    warning(paste0("No probes for gene ", gene[[4]],"(",gene[[5]],")."))
    return(NULL) 
  }
   
  if (substr(pf_meth[1, pf_chr_colname], 1, 3) != "chr") {
    pf_meth[,pf_chr_colname] = paste0("chr",pf_meth[,pf_chr_colname])
  }
  
  # get meth infos
  if (strand == "-") {
    off_set_beg = dwn_str
    off_set_end = up_str
    tss = end
  } else {
    off_set_beg = up_str
    off_set_end = dwn_str
    tss = beg
  }

  ## Compute probes associated with the gene 
  probe_idx =   rownames(pf_meth)[
    !is.na(pf_meth[[pf_pos_colname]]) & !is.na(pf_meth[[pf_chr_colname]]) &
    pf_meth[[pf_chr_colname]] == chr &
    pf_meth[[pf_pos_colname]] >= tss-up_str &
    pf_meth[[pf_pos_colname]] < tss+dwn_str
  ]    

  if (length(probe_idx) == 0) {
    warning(paste0("No probes for gene ", gene[[4]],"(",gene[[5]],")."))
    return(NULL)
  } else {
    return(probe_idx)    
  }
}





get_multiomic_data = function(gene_symbols, tcga_project, feat_indexed_probes, region_id, interaction_range=2500) {
  # # warning: feat_indexed_probes is a global variable
  s_cnv   = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_cnv.rds"))
  s_meth  = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_meth.rds"))
  s_trscr = mreadRDS(paste0("~/projects/tcga_studies/study_", tcga_project, "_trscr.rds"))
  genes_singleton = get_genes(tcga_project)
  genes = genes_singleton$genes
  if (missing(feat_indexed_probes)) {
    # params
    pf_chr_colname = "Chromosome"
    pf_pos_colname = "Start"
    if (missing(region_id)) {
      feat = genes[gene_symbols,]
      chr = feat[[1]]
      # meth pf
      if (!exists("pf_orig")) {
        pf_orig = s_meth$platform
        pf_orig = pf_orig[pf_orig[,pf_pos_colname]>0,]    
        pf_orig = pf_orig[order(pf_orig[[pf_chr_colname]],pf_orig[[pf_pos_colname]]), ]
      }  
      meth_platform = pf_orig[pf_orig[[pf_chr_colname]]%in%feat[[1]], ]
      head(meth_platform)
      probes = get_probe_names(feat, meth_platform, pf_chr_colname, pf_pos_colname, interaction_range, interaction_range) 

      # print(feat_indexed_probes)
      tss = ifelse(feat[[6]]=="+", feat[[2]], feat[[3]])
      region_id = paste0(feat[[1]], ":", tss-interaction_range, "-", tss+interaction_range)      
      feat_indexed_probes = list()
      feat_indexed_probes[[gene_symbols]] = feat_indexed_probes[[region_id]] = probes
      feat_indexed_probes
    } else {
      tmp_pf = strsplit(region_id, ":|-", perl=TRUE)[[1]]
      feat = genes[1,1:6]
      feat[[1]] = tmp_pf[[1]]
      feat[[2]] = as.numeric(tmp_pf[[2]])
      feat[[3]] = as.numeric(tmp_pf[[3]])
      feat[[4]] = "foo"
      feat[[5]] = 0
      chr = feat[[1]]
      # meth pf
      if (!exists("pf_orig")) {
        pf_orig = s_meth$platform
        pf_orig = pf_orig[pf_orig[,pf_pos_colname]>0,]    
        pf_orig = pf_orig[order(pf_orig[[pf_chr_colname]],pf_orig[[pf_pos_colname]]), ]
      }  
      meth_platform = pf_orig[pf_orig[[pf_chr_colname]]%in%feat[[1]], ]
      head(meth_platform)
      probes = get_probe_names(feat, meth_platform, pf_chr_colname, pf_pos_colname, 0, feat[[3]]-feat[[2]])       
      feat_indexed_probes = list()
      feat_indexed_probes[[region_id]] = probes
      feat_indexed_probes
    }
    # print(feat_indexed_probes)
    # print(region_id)
  }
  if (missing(gene_symbols)) {
    tmp_reg = strsplit(region_id, ":|-", perl=TRUE)[[1]]
    chr = tmp_reg[1]
    start = as.numeric(tmp_reg[2])
    end =   as.numeric(tmp_reg[3])
    # gene_symbols = rownames(genes[as.character(genes[,1])==chr & genes[,3]>=start-1000 & genes[,2] <= end+1000,1:6])
    tmp_sub_genes = genes_singleton$chrs_indexed_genes[[chr]]
    gene_symbols = rownames(tmp_sub_genes[tmp_sub_genes[,3]>=start- interaction_range & tmp_sub_genes[,2] <= end+interaction_range,1:6])
    if (length(gene_symbols) == 0) {
      return(NULL)
    }
  }
  if (missing(region_id)) {
    gene_symbol = gene_symbols[1]
    tmp_pf = get_pf_from_feat_indexed_probes(feat_indexed_probes)
    tmp_pf = tmp_pf[as.character(tmp_pf$chr)==as.character(genes[gene_symbol, 1]), ]
    tss = ifelse(genes[gene_symbol,6]=="+", genes[gene_symbol,2], genes[gene_symbol,3])
    reg = tmp_pf[tmp_pf[,2]<tss&tmp_pf[,3]>tss, ]
    region_id = paste0(reg[1], ":", reg[2], "-", reg[3])
  }
  # s_cnv = NULL
  ret = preproc_omics_data(region_id, gene_symbols, s_cnv, s_meth, s_trscr, feat_indexed_probes)
  return(ret)
}

# debugged and optimized version of dmprocr::trscr_meth_analysis https://github.com/bcm-uga/dmprocr
preproc_omics_data = function(region_id, gene_symbols, s_cnv, s_meth, s_trscr, feat_indexed_probes) {
  # meth_data
  # meth_probe_idx = intersect(feat_indexed_probes[[region_id]], rownames(s_meth$data))
  meth_probe_idx = feat_indexed_probes[[region_id]]
  if (length(meth_probe_idx) <= 1) {
      return(NULL)
  }
  tmp_reg = strsplit(region_id, ":|-", perl=TRUE)[[1]]
  chr = tmp_reg[1]
  meth_data = s_meth$data[meth_probe_idx, ]
  meth_data = meth_data[, apply(is.na(meth_data), 2, sum)/nrow(meth_data) < 0.5]
  meth_data = meth_data[apply(is.na(meth_data), 1, sum)/ncol(meth_data) < 0.5, ]
  # dim(meth_data)
  meth_probe_idx = rownames(meth_data)
  # idx_sample according to s_cnv if needed
  if (length(gene_symbols)==1) {
    tmp_gene_symbols = c(gene_symbols, gene_symbols)
  } else {
    tmp_gene_symbols = gene_symbols
  }
  gene_symbol = gene_symbols[1]
  FAST = FALSE
  if (FAST) {
    idx_sample = intersect(colnames(s_trscr$data), colnames(meth_data))
  } else {
    if (!is.null(s_cnv)) {
      idx_sample = intersect (
        intersect(
          colnames(s_trscr$data)[order(s_trscr$data[gene_symbol,])],
          colnames(meth_data)
        ),
        colnames(s_cnv$data)[apply(abs(s_cnv$data[tmp_gene_symbols, ]) < 0.2, 2, all)]
      )
    } else {
      idx_sample = intersect(
        colnames(s_trscr$data)[order(s_trscr$data[gene_symbol,])],
        colnames(meth_data)
      )
    }
    if (length(idx_sample) <= 1) {
      return(NULL)
    }      
  }
  d = data.frame(t(meth_data[, idx_sample]))
  tmp_trscr_data = data.frame(t(data.frame(s_trscr$data[tmp_gene_symbols, idx_sample])))
  genes_to_keep = sapply(gene_symbols, function(g) {
    if (length(unique(tmp_trscr_data[[g]])) == 1) {
      return(FALSE)
    } else {
      return(TRUE)
    }
  })
  gene_symbols = gene_symbols[genes_to_keep]
  for (g in gene_symbols) {
    d[[g]] = tmp_trscr_data[[g]]
  }
  ret = list(
    d=d,
    probes=meth_probe_idx,
    gene_symbols=gene_symbols,
    region_id=region_id,
    tcga_project=tcga_project
  )
  return(ret)
}


momic_pattern = function(gene_symbols, tcga_project, interaction_range=2500, LAYOUT=TRUE, col, ...) {    
  data = get_multiomic_data(gene_symbols=gene_symbols, tcga_project=tcga_project, interaction_range=interaction_range, ...)
  if (LAYOUT) {layout(matrix(c(2,1,1,1,1), 1))}
  # layout(matrix(c(1, 1, 2, 2, 2, 2), 2), respect=TRUE)
  # transcriptome
  # par(mar=c(10, 4.1, 4.1, 2.1))
  # methylome
  colors = c("cyan", "black", "red")
  cols = colorRampPalette(colors)(20)
  breaks = seq(0, 1, length.out = length(cols) + 1)
  main = paste0(tcga_project, " ", data$gene_symbol, " expression and TSS+/-", interaction_range/1000, "kb methylation")
  # par(mar=c(10, 4.1, 4.1, 2.1))

  par(mar=c(5.7, 0, 4.1, 2.1))
  image(t(data$d[,data$probes]), col=cols, breaks=breaks, xaxt="n", yaxt="n", main=main)
  axis(1, (1:nrow(t(data$d[,data$probes])) - 1)/(nrow(t(data$d[,data$probes])) - 1), rownames(t(data$d[,data$probes])), las = 2)

  par(mar=c(5.7, 4.1, 4.1, 0))
  if (missing(col)) {
    col=1
    type="l"
  } else {
    col = adjustcolor(col[rownames(data$d)], .3)
    type="p"
  }
  plot(data$d[,gene_symbols[1]], (seq(0,1,length=nrow(data$d)+1)[-1]) - .5/nrow(data$d), 
    main="", 
    xlab="expr.", 
    ylab=paste0(nrow(data$d), " samples"), 
    yaxt="n",
    ylim=c(0,1), 
    type=type,
    lwd=3,
    col=col,
    yaxs = "i"
  ) 
  par(mar=c(5.1, 4.1, 4.1, 2.1))
  return(data)
}

tcga_project = "TCGA-BRCA"
interaction_range=5000


gene_symbols = "DNMT3B" 

for (gene_symbols in gecs) {
  
  idx_normal = rownames(exp_grp)[exp_grp$tissue_status=="normal"]
  idx_normal = intersect(colnames(s_trscr$data), idx_normal)
  idx_off = rownames(exp_grp)[exp_grp[, paste0(g, "_off_on")]%in%0]
  idx_off = intersect(colnames(s_trscr$data), idx_off)
  idx_on = rownames(exp_grp)[exp_grp[, paste0(g, "_off_on")]%in%1]
  idx_on = intersect(colnames(s_trscr$data), idx_on)
  
  col = rep(1, length(c(idx_normal, idx_off, idx_on)))
  names(col) = c(idx_normal, idx_off, idx_on)
  col[idx_normal] = 3
  col[idx_off]    = 4
  col[idx_on]     = 2
  
  print(gene_symbols)
  data = momic_pattern(gene_symbols, tcga_project, interaction_range, col=col)
  pvs = sapply(data$probes, function(p) {
    m = lm(data$d[,data$gene_symbols]~data$d[,p]      )
    pv = anova(m)[1,5]
    print(pv)
    ret = pv
    ret
  })

  layout(matrix(1:3, 1), respect=TRUE)
  barplot(-log10(pvs), las=2, ylab="-log10(pv)", main=paste0("expr. ~ probe (Student)")) 
  mmeth = apply(data$d[,data$probes], 1, mean)
  m = lm(data$d[,data$gene_symbols]~  mmeth)
  pv = anova(m)[1,5]
  plot(mmeth, data$d[,data$gene_symbols], main=paste0("expr. ~ av. meth., Student pv=", signif(pv,3)), xlab="av. meth.", ylab="expr. (log2(norm_count))", col=col[names(mmeth)])
  abline(m, col=2)
  m = lm(mmeth[c(idx_off, idx_on)]~c(rep("OFF", length(idx_off)), rep("ON", length(idx_on))))
  pv_t = anova(m)[1,5]
  pv_mw = wilcox.test(mmeth[c(idx_off, idx_on)]~c(rep("OFF", length(idx_off)), rep("ON", length(idx_on))))$p.value
  boxplot(mmeth[c(idx_off, idx_on)]~c(rep("OFF", length(idx_off)), rep("ON", length(idx_on))), col=c(4,2), ylab="av. meth.", main=paste0("av. meth. ~ ON|OFF"), xlab=paste0("Student pv=", signif(pv_t,3), " Mann-Whitney pv=", signif(pv_mw,3)))
}


```


# Session Information

```{r, results="verbatim"}
sessionInfo()
```

